src/python/pants/engine/build_files.py

```diff
@@ -201,108 +202,99 @@ def _hydrate(item_type, spec_path, **kwargs):
   return item
 
 
+class _MappedSpecs(datatype([
+    ('address_families', tuple),
+    ('specs', Specs),
+])):
+  """Wraps up logic to resolve specs into individual target addresses, or error out."""
+
+  def __new__(cls, address_families, specs):
+    return super(_MappedSpecs, cls).__new__(cls, tuple(address_families), specs)
+
+  @memoized_property
+  def _exclude_compiled_regexps(self):
+    return [re.compile(pattern) for pattern in set(self.specs.exclude_patterns or [])]
+
+  def _excluded_by_pattern(self, address):
+    return any(p.search(address.spec) is not None for p in self._exclude_compiled_regexps)
```

Comment:
Nice! Very Pythonic 🐍

Reply:
This was just stolen from elsewhere in the file so it is indeed pythonic but not quite my idea.